Builds logic that manages turfs contained inside an area

LemonInTheDark opened the pull request:
## About The Pull Request

Area contents isn't a real list, instead it involves filtering
everything in world
This is slow, and something we should have better support for.

So instead, lets manage a list of turfs inside our area. This is simple,
since we already move turfs by area contents anyway

This should speed up the uses I've found, and opens us up to using this
pattern more often, which should make dev work easier.

By nature this is a tad fragile, so I've added a unit test to double
check my work

Rather then instantly removing turfs from the contained_turfs list, we
enter them into a list of turfs to pull out, later.
Then we just use a getter for contained_turfs rather then a var read

This means we don't need to generate a lot of usage off removing turf by
turf from space, and can instead do it only when we need to

I've added a subsystem to manage this process as well, to ensure we
don't get any out of memory errors. It goes entry by entry, ensuring we
get no overtime.
This allows me to keep things like space clean, while keeping high
amounts of usage on a sepearate subsystem when convienient

As a part of this goal of keeping space's churn as low as possible, I've
setup code to ensure we do not add turfs to areas during a z level
increment adjacent mapload. this saves a LOT of time, but is a tad
messy

I've expanded where we use contained_turfs, including into some cases
that filter for objects in areas. need to see if this is sane or not.

Builds sortedAreas on demand, caching until we mark the cache as violated

It's faster, and it also has the same behavior

I'm not posting speed changes cause frankly they're gonna be a bit scattered and I'm scared to.
@Mothblocks if you'd like I can look into it. I think it'll pay for itself just off `reg_in_areas_in_z` (I looked into it. it's really hard to tell, sometimes it's a bit slower (0.7), sometimes it's 2 seconds (0.5 if you use the old master figure) faster. life is pain.)

## Why It's Good For The Game

Less stupid, more flexible, more speed

LemonInTheDark committed:
Builds logic that manages turfs contained inside an area

Area contents isn't a real list, instead it involves filtering
everything in world
This is slow, and something we should have better support for.

So instead, lets manage a list of turfs inside our area. This is simple,
since we already move turfs by area contents anyway

This should speed up the uses I've found, and opens us up to using this
pattern more often, which should make dev work easier.

By nature this is a tad fragile, so I've added a unit test to double
check my work

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into potato_contents

san7890 commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -737,11 +741,25 @@ GLOBAL_LIST_EMPTY(the_station_areas)
 	// We'll have to update this if offsets change, because we load lowest z to highest z
 	generate_lighting_appearance_by_z(z_value)
 
+/datum/controller/subsystem/mapping/proc/build_area_turfs(z_level, space_guarenteed)
+	// If we know this is filled with default tiles, we can use the default area
+	// Faster
+	if(space_guarenteed)
```
```suggestion
/datum/controller/subsystem/mapping/proc/build_area_turfs(z_level, space_guaranteed)
	// If we know this is filled with default tiles, we can use the default area
	// Faster
	if(space_guaranteed)
```

spellcheck

LemonInTheDark committed:
you saw nothing

Mothblocks commented on code/__DEFINES/turfs.dm:
```diff
@@ -22,6 +22,8 @@
 ///Returns all turfs in a zlevel
 #define Z_TURFS(ZLEVEL) block(locate(1,1,ZLEVEL), locate(world.maxx, world.maxy, ZLEVEL))
 
+#define ALL_TURFS(...) block(locate(1, 1, 1), locate(world.maxx, world.maxy, world.maxz))
```
Needs doc-comment

Mothblocks commented on code/modules/unit_tests/area_contents.dm:
```diff
@@ -0,0 +1,29 @@
+/// Verifies that an area's perception of their "turfs" is correct, and no other area overlaps with them
+/// Quite slow, but needed
```
If it's slow, then put it in a later priority (probably doesn't need to be its own, just a "slow" priority is fine)

Mothblocks commented on code/modules/unit_tests/area_contents.dm:
```diff
@@ -0,0 +1,29 @@
+	var/list/turf_to_area = list()
+	// First, we check that there are no entries in more then one area
+	// That or duplicate entries
+	for(var/area/lad in world)
+		for(var/turf/thing as anything in lad.contained_turfs)
```
bro real names

Mothblocks commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -737,11 +741,25 @@ GLOBAL_LIST_EMPTY(the_station_areas)
+		global_area.contained_turfs += Z_TURFS(z_level)
+		return
+
+	for(var/turf/lad as anything in Z_TURFS(z_level))
```
real name

you are using "lad" both to mean area in one place and turf to mean other

this is just x but longer

Mothblocks reviewed (changes requested):
i think its fine as long as youre sure the memory isnt going to be a problem

how does it work with, say, user created areas?

LemonInTheDark committed:
cleans up the unit test, makes it run later, docs some stuff and applies better var names

LemonInTheDark committed:
Spelling, my mortal enemy

Thanks san

Co-authored-by: san7890 <the@san7890.com>

LemonInTheDark commented:
> i think its fine as long as youre sure the memory isnt going to be a problem
> 
> how does it work with, say, user created areas?

At 255^2 turfs per z, with a ballpark cost of 8 bytes per list entry I expect 520kb per z level, or about 6 megabtyes for our standard ~11/~13 level count

It works as expected. we just hook into the same bit of code inside `/proc/create_area` that actually moves the turf over, and additionally register it. that's what I've done for this whole pr really

LemonInTheDark commented:
Currently adds like 20 seconds to init
Fortunately, I am highly competent, so that number can go down significantly

LemonInTheDark committed:
Speeds up mapload back to what it should be, roughly

Rather then instantly removing turfs from the contained_turfs list, we
enter them into a list of turfs to pull out, later.
Then we just use a getter for contained_turfs rather then a var read

This means we don't need to generate a lot of usage off removing turf by
turf from space, and can instead do it only when we need to

I've added a subsystem to manage this process as well, to ensure we
don't get any out of memory errors. It goes entry by entry, ensuring we
get no overtime.
This allows me to keep things like space clean, while keeping high
amounts of usage on a sepearate subsystem when convienient

As a part of this goal of keeping space's churn as low as possible, I've
setup code to ensure we do not add turfs to areas during a z level
increment adjacent mapload. this saves a LOT of time, but is a tad
messy

I've expanded where we use contained_turfs, including into some cases
that filter for objects in areas. need to see if this is sane or not.

Oh also I added a global area list, and replaced all non sort caring
instances of sortedAreas with it, since it seemed kinda silly

LemonInTheDark committed:
Merge branch 'potato_contents' of https://github.com/LemonInTheDark/tgstation into potato_contents

LemonInTheDark committed:
Builds sortedAreas on demand, caching until we mark the cache as violated

It's faster, and it also has the same behavior

LemonInTheDark committed:
missed a spot

LemonInTheDark committed:
cleans some warnings, makes get_area_turfs twice as fast and less unreadable

LemonInTheDark committed:
flips a contain turfs from true to false, removes a warn

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into potato_contents

LemonInTheDark committed:
some as anything's, cleans teleport loc work

LemonInTheDark committed:
shit

LemonInTheDark commented:
alternates between 0.7 seconds slower and 2 seconds faster then master, depending on the position of the stars and the relationship between mercury and jupiter

I think it's worth it, up to you

Mothblocks commented:
I think you at the very least need to figure out why there's such a crazy disparity

Mothblocks commented:
Especially since there are more straight forward ways of reducing the cost of `reg_in_areas_in_z()` by investigating its consumers without this

LemonInTheDark commented:
The trouble is I KNOW it reduces the cost of many problem children, far more then it itself costs in terms of generation.
But it's very hard for me to profile the gain of the changes, just because of variablility in cost between tests and the scope of the change.
Sometimes master will init in 59 seconds while the changes init in 57.7, sometimes master inits in 58.2 when we init in 58.7 (I think this was the cost of ssassets being weird, something is fucky with the preferences json)

Mothblocks commented:
Please try to investigate because that seems like enough noise to make this very hard to truly weigh the costs of

msgerbs commented:
Does this have any performance improvement for anything with shuttles? I'm having a busy week so haven't had a chance to look myself.

LemonInTheDark commented:
I'm unsure. it touches on a proc in the shuttle chain, but that's on load not move.

LemonInTheDark commented:
init time comparison between potato_contents and master, on the same master commit
I'd sample a branch, checkout the other branch, and repeat alternating
the last 4 were done with no other backround programs running, steam, discord, firefox, etc. they're much more consistent
```
master:
57.9
60.8
59.8
59.4
57.9
post focusing:
56.4
56.9
58.6 
56.5
potato_contents:
58.3
56.4
62.1
59.9
60.4
post focusing:
55.5
55.6
55.7
55.5
```
it seems to save about a second of init
